bigqueue.go

```diff
@@ -0,0 +1,106 @@
+package bigqueue
+
+import (
+	"errors"
+	"fmt"
+	"path"
+)
+
+const (
+	cDataFileFmt  = "data_%d.dat"
+	cDataFileSize = 128 * 1024 * 1024
+	cInt64Size    = 8
+)
+
+var (
+	// ErrEmptyQueue is returned when peek/dequeue is performed on an empty queue
+	ErrEmptyQueue = errors.New("queue is empty")
+)
+
+// BigQueue implements IBigQueue interface using Go slices
+type BigQueue struct {
+	index     *QueueIndex
+	arenaList []*Arena
+	dir       string
+}
+
+// NewBigQueue constructs an instance of *BigQueue
+func NewBigQueue(dir string) (*BigQueue, error) {
+	index, err := NewQueueIndex(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	headAid, _ := index.GetHead()
+	tailAid, _ := index.GetTail()
+	arenaList := make([]*Arena, 0)
+
+	bq := &BigQueue{
+		index:     index,
+		arenaList: arenaList,
+		dir:       dir,
+	}
+
+	for i := headAid; i <= tailAid; i++ {
+		if err := bq.getNewArena(i); err != nil {
+			return nil, err
+		}
+	}
+
+	return bq, nil
+}
+
+// IsEmpty returns true when queue is empty
+func (bq *BigQueue) IsEmpty() bool {
+	headAid, headOffset := bq.index.GetHead()
+	tailAid, tailOffset := bq.index.GetTail()
+
+	return headAid == tailAid && headOffset == tailOffset
+}
+
+// Peek returns the head of the queue
+func (bq *BigQueue) Peek() ([]byte, error) {
+	if bq.IsEmpty() {
+		return nil, ErrEmptyQueue
+	}
+
+	aid, offset := bq.index.GetHead()
+
+	// check if length is present in same arena, if not get next arena
+	if cDataFileSize-offset < cInt64Size {
+		aid, offset = aid+1, 0
+	}
+
+	// read length
+	length := int(bq.arenaList[aid].ReadUint64(offset))
+	offset += cInt64Size
+	if offset == cDataFileSize {
+		aid, offset = aid+1, 0
+	}
+
+	// read message
+	_, _, message, err := bq.readBytes(aid, offset, length)
+	if err != nil {
+		return message, err
+	}
+
+	return message, nil
+}
+
+// getNewArena creates arena with aid and adds it to big queue arenaList
+func (bq *BigQueue) getNewArena(aid int) error {
+	file := path.Join(bq.dir, fmt.Sprintf(cDataFileFmt, aid))
+	arena, err := NewArena(file, cDataFileSize)
+	if err != nil {
+		return err
+	}
+	bq.arenaList = append(bq.arenaList, arena)
```

Comment:
This doesn't look right. `arena` should go into the `i`th index of the slice. Also, it seems that `getNewArena` doesn't necessarily create arenas, it can read data from existing arenas too.